### PR TITLE
build: ensure we aren't using GNU extensions

### LINKED ISSUE #14130: Clang: "anonymous structs are a GNU extension"
https://github.com/bitcoin/bitcoin/blob/68f3c7eb080e461cfeac37f8db7034fe507241d0/src/prevector.h#L152-L158

I'm opening an issue as its fix is too tiny to become a pull request.

Can be solved by renaming the struct to
```c
     struct indirect_contents { 
         size_type capacity; 
         char* indirect; 
     }; 
```

### PR DESCRIPTION
**Pull request by fanquake**
Since we [started using](https://github.com/bitcoin/bitcoin/pull/7165) the `ax_cxx_compile_stdcxx.m4` macro we've been passing `[noext]` to indicate that we don't want to use an extended mode, i.e GNU extensions. Speaking to Cory he clarified that the intention was to "require only vanilla c++11 and turn _off_ extension support so they would fail to compile".

However in the codebase we are currently making use of some GNU extensions. We should either remove there usage, or at least amend our CXX compiler checks. I'd prefer the former.

#### anonymous structs
```bash
./prevector.h:153:9: warning: anonymous structs are a GNU extension [-Wgnu-anonymous-struct]
        struct {
```

This is fixed in https://github.com/bitcoin/bitcoin/commit/b849212c1ec01cc8633b8cdcd390da9b1051be0d.

#### variadic macros

```bash
./undo.h:57:50: warning: must specify at least one argument for '...' parameter of variadic macro [-Wgnu-zero-variadic-macro-arguments]
            ::Unserialize(s, VARINT(nVersionDummy));
```

This is taken care of in #18087.

The `LOG_TIME_*` macros introduced in #16805 make use of a [GNU extension](https://gcc.gnu.org/onlinedocs/cpp/Variadic-Macros.html).

```bash
In file included from validation.cpp:22:
./logging/timer.h:99:99: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
    BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, ## __VA_ARGS__)
                                                                                                  ^
./logging/timer.h:99:99: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
./logging/timer.h:99:99: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
./logging/timer.h:99:99: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
./logging/timer.h:99:99: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
./logging/timer.h:101:92: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
    BCLog::Timer<std::chrono::seconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, ## __VA_ARGS__)
                                                                                           ^
6 warnings generated.
```

This is fixed in 081a0ab64eb442bc85c4d4a4d3bc2c8e97ac2a6d and 612e8e138b97fc5ad2f38847300132a8fc423c3f.

#### prevention
To ensure that usage doesn't creep back in we can add [`-Wgnu`](https://clang.llvm.org/docs/DiagnosticsReference.html#wgnu) to our compile time flags, which will make Clang warn whenever it encounters GNU extensions.

This would close #14130.
Also related to #17230, where it's suggested we use a GNU extension, the `gnu::pure` attribute.

### CODE REVIEW

**Comment by laanwj** [comment]
Concept ACK

**Comment by Empact** [comment]
Concept ACK 👍

**Comment by Empact** [comment]
Here's a set of changes to remove the `Wgnu-zero-variadic-macro-arguments` cases:
https://github.com/Empact/bitcoin/tree/2020-02-no_gnu_extensions

**Comment by fanquake** [comment]
> Here's a set of changes

Thanks, I've taken some of those changes.

**Comment by theuni** [comment]
Concept ACK, this was for sure the intent of [noext].

**Comment by practicalswift** [comment]
Concept ACK -- inside the standard is better than outside the standard

**Comment by fanquake** [comment]
Rebased on master now that #18087 is in, fixed up the sanitizer issue in prevector (bad rebase) and re-ordered some commits.

**Comment by practicalswift** [comment]
ACK 6af7c41771c2b3adeffe81ce06a70e0dbbef5360 -- diff looks correct

File: src/logging.h [hunk 1]
```diff
@@ -178,7 +178,7 @@ static inline void LogPrintf(const char* fmt, const Args&... args)
 // evaluating arguments when logging for the category is not enabled.
 #define LogPrint(category, ...)              \
     do {                                     \
-        if (LogAcceptCategory((category))) { \
+        if ((category) == BCLog::LogFlags::ALL || LogAcceptCategory((category))) { \
```
Comment (MarcoFalke) [inline src/logging.h]: Why is this change needed?

Comment (Empact) [reply src/logging.h]: Added this to ensure the FlushStateToDisk call would be logged even if no categories were active, which I believe is consistent with the current behavior - though I am not certain it is necessary.

Comment (laanwj) [reply src/logging.h]: I am not convinced at the moment that this is necessary. `Timer::Log` has the following implementation:
```
    void Log(const std::string& msg)
    {
        const std::string full_msg = this->LogMsg(msg);

        if (m_log_category == BCLog::LogFlags::ALL) {
            LogPrintf("%s\n", full_msg);
        } else {
            LogPrint(m_log_category, "%s\n", full_msg);
        }
    }
```
To me, this implies that `LogPrint(BCLog::LogFlags::ALL…` is never called. And it shouldn't, it's not valid usage.

**Comment by elichai** [comment]
Isn't `__COUNTER__` a GNU extension too?
https://renenyffenegger.ch/notes/development/languages/C-C-plus-plus/preprocessor/macros/predefined/__COUNTER__
https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html

saw it here again: https://github.com/bitcoin/bitcoin/pull/18234/commits/3cb26649716b2460dae44404f7ce2bd4715e0646#diff-998357bc40c89d644815a6fd52c99c9fR213

**Comment by fanquake** [comment]
> Isn't __COUNTER__ a GNU extension too?

From what I can find, yes `__COUNTER__` (added to GCC in [4.3](https://gcc.gnu.org/gcc-4.3/changes.html)) is an extension, however it's supported by [GCC](https://gcc.gnu.org/onlinedocs/cpp/Common-Predefined-Macros.html),  [Clang](https://clang.llvm.org/docs/LanguageExtensions.html), [MSVC](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2019) and `-Wgnu` doesn't currently warn about it. Maybe someone can address this in a followup if they'd like.

I've reverted the changes in [`logging.h`](https://github.com/bitcoin/bitcoin/pull/18088#discussion_r377616141) and did some testing to check that the "FlushStateToDisk: write coins cache to disk" call was always being logged, regardless of logging categories/options.

File: src/logging.h (hunk 1 above)
Comment (fanquake) [reply src/logging.h]: I've removed this change.

File: src/validation.cpp [hunk 2]
```diff
@@ -2327,7 +2327,7 @@ bool CChainState::FlushStateToDisk(
         // Flush best chain related state. This can only be done if the blocks / block index write was also done.
         if (fDoFullFlush && !CoinsTip().GetBestBlock().IsNull()) {
             LOG_TIME_SECONDS(strprintf("write coins cache to disk (%d coins, %.2fkB)",
-                coins_count, coins_mem_usage / 1000));
+                coins_count, coins_mem_usage / 1000), BCLog::ALL);
```
Comment (MarcoFalke) [inline src/validation.cpp]: I don't really like leaking this private implementation detail of the log flags into the caller code. I'd slightly prefer a type-safe way, that also better documents what is happening.

What about making this argument optional not with VA_ARGS, but with an `Optional<BCLog::LogFlags>`? Passing a nullopt or `{}` will unconditionally log, whereas passing a log flag will only log to that category (if enabled).

Comment (Empact) [reply src/validation.cpp]: We could drop it altogether as this is the only call to `LOG_TIME_SECONDS`:

```diff
diff --git a/src/logging.h b/src/logging.h
index 5d8ec33392..b2fde1b9ea 100644
--- a/src/logging.h
+++ b/src/logging.h
@@ -178,7 +178,7 @@ static inline void LogPrintf(const char* fmt, const Args&... args)
 // evaluating arguments when logging for the category is not enabled.
 #define LogPrint(category, ...)              \
     do {                                     \
-        if ((category) == BCLog::LogFlags::ALL || LogAcceptCategory((category))) { \
+        if (LogAcceptCategory((category))) { \
             LogPrintf(__VA_ARGS__);          \
         }                                    \
     } while (0)
diff --git a/src/logging/timer.h b/src/logging/timer.h
index 34a6e25185..6a40dc48ba 100644
--- a/src/logging/timer.h
+++ b/src/logging/timer.h
@@ -95,8 +95,8 @@ private:
 
 #define LOG_TIME_MILLIS(end_msg, log_category) \
     BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category)
-#define LOG_TIME_SECONDS(end_msg, log_category) \
-    BCLog::Timer<std::chrono::seconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category)
+#define LOG_TIME_SECONDS(end_msg) \
+    BCLog::Timer<std::chrono::seconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg)
 
 
 #endif // BITCOIN_LOGGING_TIMER_H
diff --git a/src/validation.cpp b/src/validation.cpp
index a0ac003029..bab04b8e34 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -2328,7 +2328,7 @@ bool CChainState::FlushStateToDisk(
         // Flush best chain related state. This can only be done if the blocks / block index write was also done.
         if (fDoFullFlush && !CoinsTip().GetBestBlock().IsNull()) {
             LOG_TIME_SECONDS(strprintf("write coins cache to disk (%d coins, %.2fkB)",
-                coins_count, coins_mem_usage / 1000), BCLog::ALL);
+                coins_count, coins_mem_usage / 1000));
 
             // Typical Coin structures on disk are around 48 bytes in size.
             // Pushing a new one to the database can cause it to be written
```

Comment (Empact) [reply src/validation.cpp]: Updated this branch: https://github.com/fanquake/bitcoin/compare/no_gnu_extensions...Empact:2020-02-no_gnu_extensions

Comment (fanquake) [reply src/validation.cpp]: > We could drop it altogether

I've updated to do this.

File: src/logging/timer.h [hunk 3]
```diff
@@ -93,12 +93,10 @@ class Timer
...
-#define LOG_TIME_MICROS(end_msg, ...) \
-    BCLog::Timer<std::chrono::microseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, ## __VA_ARGS__)
-#define LOG_TIME_MILLIS(end_msg, ...) \
-    BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, ## __VA_ARGS__)
-#define LOG_TIME_SECONDS(end_msg, ...) \
-    BCLog::Timer<std::chrono::seconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, ## __VA_ARGS__)
+#define LOG_TIME_MILLIS(end_msg, log_category) \
+    BCLog::Timer<std::chrono::milliseconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg, log_category)
+#define LOG_TIME_SECONDS(end_msg) \
+    BCLog::Timer<std::chrono::seconds> PASTE2(logging_timer, __COUNTER__)(__func__, end_msg)
```
Comment (MarcoFalke) [inline src/logging/timer.h]: So while the macro name indicates the only difference is the precision, they are functionally also different to the logging behaviour: One always logs, the other must accept a category. Not sure if it makes sense to keep the names so similar then.

Comment (fanquake) [reply src/logging/timer.h:99]: Ok. I've renamed `LOG_TIME_MILLIS` to `LOG_TIME_MILLIS_WITH_CATEGORY`.

**Comment by fanquake** [comment]
Rebased after #18591.

**Comment by vasild** [comment]
utACK 0ae8f18df

If we currently do not see any warnings due to the newly introduced `-Wgnu` then I would suggest to also add `-Werror=gnu` (if compiling with `--enable-werror`). So that a newly introduced warning in the future does not sneak - printed, but unnoticed in CI logs.

**Comment by dongcarl** [comment]
ACK 0ae8f18dfe143051fec6ae10ea7df10142e3ff2f
-----
Read relevant gcc docs, built, and ran tests.

**Comment by practicalswift** [comment]
ACK 0ae8f18dfe143051fec6ae10ea7df10142e3ff2f -- diff looks correct

Agree with @vasild about `-Werror=gnu` as part of `--enable-werror`. May also be done in a follow-up.

**Comment by MarcoFalke** [comment]
ACK 0ae8f18dfe143051fec6ae10ea7df10142e3ff2f

**Comment by practicalswift** [comment]
@vasild Would you mind taking on the `-Werror=gnu` thing? :)

**Comment by vasild** [comment]
> @vasild Would you mind taking on the `-Werror=gnu` thing? :)

Here it is: https://github.com/bitcoin/bitcoin/pull/18887 build: enable -Werror=gnu
